File: Day 05/Python/sol.py
"""
FILE: Day 05/Python/sol.py
Author: Alex Jones
Desc: Solution to day 5 problems (9 & 10) for Advent of Code 2024, solved in Python 3.
"""
data = open("Day 05/data.txt", "r").read().strip().replace("\r","").split("\n\n")
orderings = [tuple(x.split("|")) for x in data[0].split("\n")]
updates = [x.split(",") for x in data[1].split("\n")]

from collections import defaultdict
from functools import cmp_to_key

comes_before = defaultdict(set)
for (pred, succ) in orderings:
    comes_before[pred].add(succ)
p1, p2 = 0, 0
for update in updates:
    ordered = sorted(update, key=cmp_to_key(lambda fst, snd: -1 * (snd in comes_before[fst])))
    p1 += int(update[len(update) // 2]) if update == ordered else 0
    p2 += int(ordered[len(ordered) // 2]) if update != ordered else 0
print("Problem 9:", p1)
print("Problem 10:", p2)

File: Day 05/Rust/sol.rs
/*
 * FILE: Day 05/Rust/sol.rs
 * Author: Alex Jones
 * Desc: Solution to day 5 problems (9 & 10) for Advent of Code 2024, solved in Rust.
 */
use std::fs;
use std::collections::HashMap;
use std::cmp::Ordering;

pub fn solve() {
    let contents: String = fs::read_to_string("Day 05/data.txt")
        .expect("Could not read data file")
        .trim()
        .replace("\r", "");
    let data: Vec<&str> = contents
        .split("\n\n")
        .collect();
    let orderings = data[0].split("\n")
        .map(|x| {
            let elems = x.split("|").collect::<Vec<_>>();
            (elems[0], elems[1])
        })
        .collect::<Vec<_>>();
    let updates = data[1].split("\n")
        .map(|x| x.split(",").collect::<Vec<_>>())
        .collect::<Vec<_>>();

    let mut comes_before: HashMap<&str,Vec<&str>> = HashMap::new();
    for (pred, succ) in orderings {
        if comes_before.contains_key(pred) {
            comes_before.get_mut(pred).unwrap().push(succ);
        } else {
            comes_before.insert(pred, vec![succ]);
        }
    }

    let (mut p1, mut p2) = (0, 0);
    for update in updates {
        let mut ordered = update.clone();
        ordered.sort_by(|fst, snd| {
            if comes_before[fst].contains(snd) { Ordering::Less }
            else { Ordering::Equal }
        });
        if update == ordered {
            p1 += update[update.len()/2].parse::<i32>().unwrap();
        } else {
            p2 += ordered[ordered.len()/2].parse::<i32>().unwrap();
        }
    }
    
    println!("Problem 9: {}", p1);
    println!("Problem 10: {}", p2);
}